fobnail-attester.c

```diff
@@ -0,0 +1,158 @@
+					   -1,
+					   0,
+					   res_buf_len,
+					   res_buf,
+					   NULL,
+					   res_buf);
+	if (ret != 0)
+		fprintf(stderr, "Err: cannot response.\n");
+
+}
+
+void att_coap_add_resource(struct coap_context_t* coap_context,
+			   coap_request_t method, const char* resource_name,
+			   coap_method_handler_t handler)
+{
+        coap_str_const_t* resource_uri = coap_new_str_const((uint8_t const*)resource_name, strlen(resource_name));
+        coap_resource_t* resource = coap_resource_init(resource_uri, COAP_RESOURCE_FLAGS_RELEASE_URI);
+        coap_register_handler(resource, method, handler);
+        coap_add_resource(coap_context, resource);
+}
+
+coap_endpoint_t* att_coap_new_endpoint(coap_context_t* coap_context,
+        			       const char* listen_address, const uint16_t port,
+        			       const coap_proto_t coap_protocol)
+{
+        /* prepare address */
+        coap_address_t addr = {0};
+        coap_address_init(&addr);
+        addr.addr.sin.sin_family = AF_INET;
+        inet_pton(AF_INET, listen_address, &addr.addr.sin.sin_addr);
+        addr.addr.sin.sin_port = htons(port);
+
+        /* create endpoint */
+        return coap_new_endpoint(coap_context, &addr, coap_protocol);
+}
+
+coap_context_t* att_coap_new_context(const bool enable_coap_block_mode) {
+        /* startup */
+        coap_startup();
+
+        /* create new context */
+        coap_context_t* coap_context = NULL;
+        if ((coap_context = coap_new_context(NULL)) != NULL) {
+                if (enable_coap_block_mode) {
+                        /* enable block handling by libcoap */
+                        coap_context_set_block_mode(coap_context, COAP_BLOCK_USE_LIBCOAP | COAP_BLOCK_SINGLE_BODY);
+                }
+        }
+
+        return coap_context;
+}
+
+/* --------------------------- main ----------------------------- */
+
+int main(int argc, char *argv[])
+{
+	int result;
+
+	/* signal handling */
+	signal(SIGINT, handle_sigint);
+
+	/* TODO: parse CLI arguments if needed */
+	/*result = parse_command_line_arguments(argc, argv, ...) != 0
+	if (result) {
+		// 1 means help message is displayed, -1 means error
+		return (result == 1) ? EXIT_SUCCESS : EXIT_FAILURE;
+	}*/
+
+	/* TODO: generate attestation identity key */
+
+	/* TODO: sign attestation identity key */
+
+	coap_context_t* coap_context = NULL;
+	coap_endpoint_t* coap_endpoint = NULL;
+
+	coap_context = att_coap_new_context(true);
+	if (coap_context == NULL) {
+		printf("Cannot create CoAP context.\n");
+		goto error;
+	}
+
+	printf("Creating CoAP server endpoint using UDP.\n");
+	if ((coap_endpoint = att_coap_new_endpoint(coap_context, LISTEN_ADDRESS, port, COAP_PROTO_UDP)) == NULL) {
+		printf("Cannot create CoAP server endpoint based on UDP.\n");
+		goto error;
+	}
+
+	/* register CoAP resource and resource handler */
+	printf("Registering CoAP resources.\n");
+	att_coap_add_resource(coap_context, COAP_REQUEST_FETCH, "attest", coap_attest_handler);
```

Comment:
Hm... I  tried with GET request and didn't get any response in my case..

Reply:
@0xDen It could be caused by the missing call to `coap_pdu_set_code`. For me both requests work. 
At first I thought FETCH is not supported in CoAP library we use with Rust, but I just grepped in wrong directory for it.
Let's stick with FETCH method as it provides the advantage that we can send request body, this may be helpful in future.

Reply:
@arturkow2000  Ok. sounds reasonable.